### PR TITLE
kvstore push row sparse

### PR DESCRIPTION
**Pull request by reminisce**
Unit test passes. Will do benchmark after refactoring done.
@eric-haibin-lin Please review.

### CODE REVIEW

File: src/common/utils.h [hunk 1]
```diff
@@ -169,6 +170,57 @@ inline int GetExecNumMatchColor() {
...
+ * the comparator comp.
+ * If the length of the range [first, last) is greater than a certain threshold,
+ * the range will be recursively divided into two and assign two threads
+ * to sort each half range.
+ * Ref: https://github.com/dmlc/difacto/blob/master/src/common/parallel_sort.h
+ */
+template<typename RandomIt, typename Compare>
+void ParallelSort(RandomIt first, RandomIt last, size_t num_threads, Compare comp) {
+  const auto num = std::distance(first, last);
+  size_t grainsize = std::max(num / num_threads + 5, static_cast<size_t>(1024*16));
```
Comment (eric-haibin-lin) [inline src/common/utils.h]: where does `num / num_threads + 5 ` come from?

File: src/kvstore/comm.h [hunk 2]
```diff
@@ -168,7 +174,7 @@ class CommCPU : public Comm {
 
   // serial implementation of reduce sum for row sparse NDArray.
   // TODO(haibin) use openmp kernel to parallelize the summation
```
Comment (eric-haibin-lin) [inline src/kvstore/comm.h]: remember to remove the todo here in next PR.

File: src/kvstore/comm.h [hunk 3]
```diff
@@ -239,6 +245,115 @@ class CommCPU : public Comm {
...
+        auto out_values = out->data().FlatTo2D<cpu, DType>();
+        auto out_indices = out->aux_data(rowsparse::kIdx).FlatTo1D<cpu, IType>();
+        for (size_t i = row_block_start; i < row_block_end; ++i) {
+          out_indices[i] = uniq_row_idx[i];
+        }
+        for (const auto& nd : nds) {
+          if (nd.storage_initialized()) {
+            const auto nd_indices = nd.aux_data(rowsparse::kIdx).FlatTo1D<cpu, IType>();
+            const auto nd_values = nd.data().FlatTo2D<cpu, DType>();
+            const auto nd_num_rows = nd.aux_shape(rowsparse::kIdx).Size();
```
Comment (eric-haibin-lin) [inline src/kvstore/comm.h]: Another thing we need to double check is the use of .Size() in our code - which returns a `size_t`. I know that the csr_to_dense function doesn't deal with really large matrix (number of element > 4 billion), that's why Jiajie's test only tested up to 20M features with k = 200. Change the interface of `Size()` will lead to many compiler errors, we might want to discuss if we want to do it or not.

File: src/kvstore/comm.h [hunk 4]
```diff
@@ -239,6 +245,115 @@ class CommCPU : public Comm {
...
+            // skip this nd if all of its row indices are smaller than out_indices[row_block_start]
+            // or current row block is not covered by [*row_idx_ptr, nd_indices_end).
+            if (nd_indices_end == row_idx_ptr || *row_idx_ptr > out_indices[row_block_end-1]) {
+              continue;
+            }
+            for (size_t irow = row_block_start;
+                 irow < row_block_end && row_idx_ptr != nd_indices_end;) {
+              if (out_indices[irow] == *row_idx_ptr) {
+                auto out_value_cur_row = out_values[irow];
+                const auto offset = std::distance(nd_indices_start, row_idx_ptr);
```
Comment (eric-haibin-lin) [inline src/kvstore/comm.h]: Is a function call gonna be overkill? equivalent to `row_idx_ptr - nd_indices_start`?

File: src/common/utils.h (hunk 1 above)
Comment (reminisce) [inline src/common/utils.h]: From Mu's code. It's can be seen under ps-lite/include/ps/internal/parallel_sort.h. 5 and 1024*16 are two magic numbers I'm not sure of.

File: src/kvstore/comm.h (hunk 2 above)
Comment (reminisce) [inline src/kvstore/comm.h]: I will remove it now.

File: src/kvstore/comm.h (hunk 4 above)
Comment (reminisce) [inline src/kvstore/comm.h]: Yah, it's an overkill. I was thinking about making the code piece generic enough to deal with any types of iterator/pointer. But realized later that the pointer type is highly unlikely to be changed in the future. I will replace all the std::distance with direct subtraction between two iterators/pointers.

File: src/common/utils.h (hunk 1 above)
Comment (eric-haibin-lin) [inline src/common/utils.h]: not sure if `src/common/utils.h` is the best place to put the code. Maybe better to put it under `src/operator`? Could be useful for other operators

Comment (reminisce) [inline src/common/utils.h]: I think this function is more widely applicable to than just operators. Putting it under src/operator makes it look like only for operators. But the fact is that any components under src/ could use it. For example src/kvstore uses this function.

**Comment by eric-haibin-lin** [comment]
test failed: 
https://ci.appveyor.com/project/piiswrong/mxnet/build/1.0.3931

```
         c:\projects\mxnet\src\kvstore\./comm.h(327): error C3016: 'i' : index variable in OpenMP 'for' statement must have signed integral type (C:\projects\mxnet\src\kvstore\kvstore.cc) [C:\projects\mxnet\build\mxnet.vcxproj]
```

**Comment by reminisce** [comment]
@eric-haibin-lin 

- Compile settings: `USE_BLAS=openblas, USE_CUDA=0, USE_OPENMP=1`
- KVStore setup: `number of cpu devices = 4`
- Use env variable `MXNET_KVSTORE_SERIAL_PUSH=1 or 0` to switch between serial and parallel push
- Benchmark results of kvstore.push row sparse matrices using serial and parallel element-wise sum


| num_rows | num_cols | nnr | density | serial(ms)  | parallel (ms)    | serial/parallel |
|:----------:|-----------|:--:|:--------:|:-----------:|:---------------:|:---------------:|
| 50,000             | 50      | 1,969| 0.01 |    3.002        |     0.810              |       3.7|
| 50,000             | 50      | 9,193| 0.05 |    36.55        |     1.655              |       22.1|
| 50,000             | 50      | 17,263| 0.1 |    140.96        |     2.996              |      47.1|
| 50,000             | 50      | 38,041| 0.3 |    1003.14        |     5.594              |       179.3|

--------------------------------------------------------------------------------------------------------

- Compile settings: `USE_BLAS=openblas, USE_CUDA=1, USE_OPENMP=1`
- KVStore setup: `number of cpu devices = 4`
- Use env variable `MXNET_KVSTORE_SERIAL_PUSH=1 or 0` to switch between serial and parallel push
- Benchmark results of kvstore.push row sparse matrices using serial and parallel element-wise sum


| num_rows | num_cols | nnr | density | serial(ms)  | parallel (ms)    | serial/parallel |
|:----------:|-----------|:--:|:--------:|:-----------:|:---------------:|:---------------:|
| 50,000             | 50      | 2,029| 0.01 |    1.096        |     1.385              |       0.79 |
| 50,000             | 50      | 9,093| 0.05 |    2.832        |     4.873              |       0.58|
| 50,000             | 50      | 17,069| 0.1 |    4.456        |     7.514              |      0.59|
| 50,000             | 50      | 38,056| 0.3 |    10.197        |     12.001              |       0.85|

-----------------------------------------------------------------------------------------------------
**Question: Why `serial push` compiled with `USE_CUDA=1` is much faster than compiled with `USE_CUDA=0`?**

**Pull request merged**